### PR TITLE
Support \( and fix instant mode switching for $, \(, \text

### LINKED ISSUE #1212: \text accepts nested $...$ fragments but not \(...\)
In LaTeX and MathJax, this works fine:
```
\[
  x = z \quad \text{because $x = y$ and \(y = z\)}
\]
```
KaTeX doesn't like the nested `\(...\)` fragment, though.

This is [documented](https://khan.github.io/KaTeX/function-support.html) and probably well-known, but I didn't find an issue on it.

This could probably be generalized to accepting only the delimiters passed in to `renderMathInElement` with `display: false`, but I don't know if the structure of the code makes that easy.

### PR DESCRIPTION
**Pull request by edemaine**
* Fix #1212 by supporting `\(...\)` in addition to `$...$` inline math nested inside `\text`. This turned out to be harder than one would think because of the way `$` was being handled specially, while `\(` needed to be handled as a function, which immediately consumed the `\(` token, which meant that the first following token got consumed in text mode instead of math mode (unlike `$` where we could switch the mode before consumption).
* Added new `consumeMode` setting for functions which tells the parser to switch modes just for the consumption of the command token.
* Now that we're working with functions, move all the `$` handling code into `functions/math.js`.  Somewhat bizarrely, we can define `$` as a function (nearly identical to `\(`) and it all works (just like we can have single-character macros).  This move removed a lot of messy stuff from `Parser.js`: `ParsedDollar`, `ParsedFuncOrArgOrDollar`, `newDollar`, `assertFuncOrArg`, and at least three explicit checks for `if (... === "$")`.
* Moved the `consume()` for the command token from `parseSymbol` to `parseGivenFunction`.  This ended up not being strictly necessary, but seems conceptually cleaner.
* Partially address #1027 by setting `consumeMode: "text"` for `\text` commands.  As a result, `\text` now instantly switches the mode to `"text"`, so even an unbraced macro argument will properly detect text mode.  I added a test for this, which is a slightly weaker form of #1027.

### CODE REVIEW

File: test/katex-spec.js [hunk 1]
```diff
@@ -872,7 +871,40 @@ describe("A text parser", function() {
...
+    });
+
+    it("should detect unbalanced $", function() {
+        expect("$").toNotParse();
+        expect("\\text{$}").toNotParse();
+    });
+
+    it("should not mix $ and \\(..\\)", function() {
+        expect("$x\\)").toNotParse();
+        expect("\\(x$").toNotParse();
```
Comment (ylemkimon) [inline test/katex-spec.js]: Shouldn't be these two in the text mode like:
```js
        expect("\\text{$x\\)}").toNotParse();
        expect("\\text{\\(x$}").toNotParse();
```
?

File: src/Parser.js [hunk 2]
```diff
@@ -964,8 +937,6 @@ export default class Parser {
                     body: arg,
                     star: star,
                 }, "text"), nucleus);
-        } else if (text === "$") {
-            return newDollar(nucleus);
```
Comment (ylemkimon) [inline src/Parser.js]: ❤️ that they are gone and became so simple!

File: src/functions/math.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
+// @flow
+import defineFunction from "../defineFunction";
+import ParseError from "../ParseError";
+
+// Switching from text mode back to math mode
+defineFunction({
+    names: ["\\(", "$"],
```
Comment (ylemkimon) [inline src/functions/math.js]: 👍

File: test/katex-spec.js (hunk 1 above)
Comment (edemaine) [reply test/katex-spec.js]: Yes, thanks for catching this!  (Otherwise we were just testing use of `$`/`\(` outside math mode again.)  Fixed.

**Comment by edemaine** [comment]
@ylemkimon Thanks for the quick review! Should be good to go now.

File: src/functions/math.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        // switching modes back.  So don't consume within expect.
+        parser.expect(close, false);
+        parser.switchMode(outerMode);
+        parser.consume();
+        return {
+            type: "styling",
+            style: "text",
+            value: body,
+        };
+    },
```
Comment (kevinbarabash) [inline src/functions/math.js]: This isn't too bad and seems to follow the same pattern as `\left`/`\right` handling.

**Pull request merged**